### DISCUSSION TITLE
How to reduce time cost of initializing containers?

### DISCUSSION THREAD

**Discussion by JungHanter** [community/community#25975] [Actions]
<p>Here is my action progress.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/2166d45ae3090d59f75fe84ecbb10b8fff2ab8ea.png" data-download-href="/uploads/short-url/4Lu4cnP1sf4iq2mIih8iAkh4epQ.png?dl=1" title="Screen Shot 2020-07-07 at 3.45.43 PM" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/1428405/181094393-fa25fbe1-f655-43f8-8a9e-bca9a9b9d87e.png" alt="Screen Shot 2020-07-07 at 3.45.43 PM" data-base62-sha1="4Lu4cnP1sf4iq2mIih8iAkh4epQ" width="690" height="384" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/2/2166d45ae3090d59f75fe84ecbb10b8fff2ab8ea_2_690x384.png, https://aws1.discourse-cdn.com/github/optimized/2X/2/2166d45ae3090d59f75fe84ecbb10b8fff2ab8ea_2_1035x576.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/2/2166d45ae3090d59f75fe84ecbb10b8fff2ab8ea.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/2166d45ae3090d59f75fe84ecbb10b8fff2ab8ea_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-07-07 at 3.45.43 PM</span><span class="informations">1358×756 45 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You can see the time of  initializing containers is very long.</p>
<p>On circle CI, the time is about 30 seconds. However in github actions, it costs twice.</p>
<p>How to reduce the time cost? (i.e. local Docker image, caching image, …)</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/junghanter">@JungHanter</a>,</p>
<p>Generally, the “<strong>Initialize containers</strong>” step mainly does the following things:</p>
<ul>
<li>Pull the images (<strong>docker pull</strong>) with the specified tags to the runner machine.</li>
<li>Create the containers (<strong>docker create</strong>) and setup the required environments.</li>
<li>Startup the containers (<strong>docker start</strong>).</li>
</ul>
<p>The “<strong>Initialize containers</strong>” step is automatically generated by GitHub Actions, and we have no methods to reduce the time it takes.</p>

**Reply by ulgens** [reply]
What about parallelization? Actions still downloads and validates service images one at a time.

**Comment by JungHanter** [comment]
<p>Thanks :slight_smile:</p>

**Comment by unaimillan** [comment]
That is sad(

![image](https://github.com/community/community/assets/12885904/6ba2bb41-c610-40cf-9508-68a15a536a5d)

Would be cool to have something like cache but for docker images 🤷

**Comment by abirchfieldSFDC** [comment]
Is it still not possible to speed up the initializing containers step?

**Reply by randall-coding** [reply]
You can if you are self hosting your runners.  

Assuming you are already doing that, you would then use a customized command script https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/customizing-the-containers-used-by-jobs#triggering-the-customization-script 

Specifically you would modify the portion of `index.js` on the `prepare_job` step, then point to the script with ACTIONS_RUNNER_CONTAINER_HOOKS.

**Comment by CpuID** [comment]
Would be awesome if these steps could parallelized on the runner side (eg. with a flag in the workflow yaml)... I suspect it may not be CPU/IO bound some of the time.

**Comment by quentindemetz** [comment]
From my experience it's the extraction step which is unbelievably slow, because layers can only be extracted in order (this is the fault of OverlayFS?) and each individual layer extraction is slow (probably single-process and CPU-bound)

**Reply by anarazel** [reply]
That's what I've also seen. It's notable that docker is a *lot* slower at that than podman at extracting containers:
https://github.com/anarazel/ci-test/actions/runs/26905240545/job/81285585269  (docker 25s, podman 14s, I've seen this across a number of runs).
